### PR TITLE
fix: improve native balance fetch logic on the UX

### LINKED ISSUE #6101: Core: Improve of native balance of the ERC-20 tokens
### **Description**
This issue aims to optimize the `TokenBalancesController` on the assets-controller.

When we change of network or account, the native balance of the ERC-20 tokens is not reset. It shows a wrong value for a couple seconds and then updates.

We should reset the state and show a skeleton text instead of a wrong value.

### **Technical Details**
- https://recordit.co/E1TCSnJeM4
- Split this ticket with a patch and follow up with a non-blocking core PR

### **Acceptance Criteria**
- Be able to show consistently the right native balance to the user

Scenario: User detect tokens on mainnet
* GIVEN a user have 15 tokens to detect
* WHEN a user detect that tokens
* THEN switch network to polygon
* AND detect tokens
* THEN switch network
* THEN token native balance should always reset and appear with the right value

Scenario: User manually imports token
* GIVEN a user go to import token screen
* WHEN a user import an ERC-20 Token
* THEN the native balance should not take longer to appear

### **References**

-   References go here.
-   Issue numbers. Links.
-   Slack threads.
-   Etc.

### PR DESCRIPTION
**Pull request by salimtb**
## **Description**

Problem: When we change network or account, the native balance of the ERC-20 tokens is not reset. It shows a wrong value for a couple seconds and then updates.

We should reset the state and show a skeleton text instead of a wrong value.

[PR](https://github.com/MetaMask/core/pull/3586) already Merged on core 

## **Related issues**

Fixes: #6101 

## **Manual testing steps**

1. Go to any network where you hold native tokens
2. switch to another network when you have 0 balance
3. you can see that the wrong balance was displayed for a short time

### **Before**

https://github.com/MetaMask/metamask-mobile/assets/26223211/d96b474f-60fc-4c4a-a43f-fde18c4a4a26

### **After**

https://github.com/MetaMask/metamask-mobile/assets/26223211/462d8c0e-dcf0-4da7-ae2f-7fdba65b9076

## **Pre-merge author checklist**

- [x] I’ve followed [MetaMask Coding Standards](https://github.com/MetaMask/metamask-mobile/blob/main/.github/guidelines/CODING_GUIDELINES.md).
- [x] I've clearly explained what problem this PR is solving and how it is solved.
- [x] I've linked related issues
- [x] I've included manual testing steps
- [x] I've included screenshots/recordings if applicable
- [x] I’ve included tests if applicable
- [ ] I’ve documented my code using [JSDoc](https://jsdoc.app/) format if applicable
- [x] I’ve applied the right labels on the PR (see [labeling guidelines](https://github.com/MetaMask/metamask-mobile/blob/main/.github/guidelines/LABELING_GUIDELINES.md)). Not required for external contributors.
- [x] I’ve properly set the pull request status:
  - [ ] In case it's not yet "ready for review", I've set it to "draft".
  - [x] In case it's "ready for review", I've changed it from "draft" to "non-draft".

## **Pre-merge reviewer checklist**

- [ ] I've manually tested the PR (e.g. pull and build branch, run the app, test code being changed).
- [ ] I confirm that this PR addresses all acceptance criteria described in the ticket it closes and includes the necessary testing evidence such as recordings and or screenshots.

### CODE REVIEW

File: patches/@metamask+assets-controllers+7.0.0.patch [hunk 1]
```diff
@@ -1,82 +1,155 @@
...
+ import { PreferencesState } from '@metamask/preferences-controller';
++import type { NetworkState } from '@metamask/network-controller';
+ /**
+  * @type AccountInformation
+  *
+@@ -29,6 +30,9 @@ export interface AccountTrackerState extends BaseState {
+     accounts: {
+         [address: string]: AccountInformation;
+     };
++    accountsByChainId: Record<string, {
```
[the remark this answers is not in the document]
Comment (salimtb) [reply patches/@metamask+assets-controllers+7.0.0.patch:16]: this patch contains only the state change in core (the addition of the `accountsByChainId` property).
the PR in core also contains other things like the provider change ( with the use of `@metamask/eth-query` ) and the use of the new polling controller ( `StaticIntervalPollingControllerV1` ) which are not yet implemented in the version used in core ( i have to create module mentioned from scratch ).
@tommasini bumped the assets controller version , but my patch is still needed to fix this bug

**Comment by chrisleewilcox** [comment]
Error on regression tests `./e2e/specs/networks/add-custom-rpc.spec.js`
image

**Review by chrisleewilcox — CHANGES_REQUESTED** [review]
https://app.bitrise.io/build/10e18cd4-22e2-4688-b017-44cb87f8716c to address test failure.

**Comment by salimtb** [comment]
> https://app.bitrise.io/build/10e18cd4-22e2-4688-b017-44cb87f8716c to address test failure.

fixed

File: app/components/Views/Wallet/index.tsx [hunk 2]
```diff
@@ -88,10 +93,12 @@ const Wallet = ({ navigation }: any) => {
   const theme = useTheme();
   const styles = createStyles(theme);
   const { colors } = theme;
+
   /**
-   * Map of accounts to information objects including balances
+   * Map of accountsByChain to information objects including balances
```
Comment (Cal-L) [inline app/components/Views/Wallet/index.tsx:98]: Seems like there's a typo here. Should it be simply Map of accounts by chainID?

**Review by Cal-L — COMMENTED** [review]
Left a comment

File: scripts/ios/bundle-js-and-sentry-upload.sh
Comment (Cal-L) [inline scripts/ios/bundle-js-and-sentry-upload.sh:1]: This should disappear once you merge with main. LMK if it doesn't

File: app/components/Views/Settings/SecuritySettings/Sections/RevealPrivateKey/RevealPrivateKey.tsx [hunk 3]
```diff
@@ -32,6 +32,7 @@ const RevealPrivateKey = () => {
   const navigation = useNavigation();
 
   const accounts = useSelector(selectAccounts);
+
```
Comment (Cal-L) [inline app/components/Views/Settings/SecuritySettings/Sections/RevealPrivateKey/RevealPrivateKey.tsx:35]: Nit - would just remove this

**Review by chrisleewilcox — APPROVED** [review]
LGTM

**Pull request merged**